### PR TITLE
Add annotation support to logging.

### PR DESCRIPTION
**Pull request by mattstep**

### CODE REVIEW

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java [hunk 1]
```diff
@@ -60,8 +60,15 @@ public StaticFormatter()
...
         return new StringWriter()
-                .append("System Context Summary")
-                .append("\t")
+                .append(TIMESTAMP_FORMATTER.format(ZonedDateTime.now()))
+                .append('\t')
+                .append(Level.INFO.name())
+                .append('\t')
+                .append(Thread.currentThread().getName())
+                .append('\t')
+                .append(StaticFormatter.class.getCanonicalName())
```
Comment (dain) [inline log-manager/src/main/java/io/airlift/log/StaticFormatter.java:69]: You could just use "SystemContext" for the logger name, and then you wouldn't have to repeat it in the log messsage.  This is just an idea... not sure if it is better or worse.

File: log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java [hunk 2]
```diff
@@ -155,4 +156,16 @@ public LoggingConfiguration setLegacyLoggerImplementationEnabled(boolean enabled
         this.legacyLoggerImplementationEnabled = enabled;
         return this;
     }
+
+    public String getSystemContextFile()
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java:160]: What's a "system context"?

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java (hunk 1 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/StaticFormatter.java:69]: That makes sense

File: log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java (hunk 2 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java:160]: It's a file that we can pass values in from kubernetes, either through the helm chart or environment variables in the container, that we add onto our logs so we can eliminate the need for fluentbit or fluentd to instrument the logs with data we know when we start the application. In particular, things like service name, pod name, ip address, etc. are valuable for debugging.

File: docs/add_logging.md [hunk 3]
```diff
@@ -38,3 +38,45 @@ public class MyClass
...
+
+* `log.output-file` can be set to the desired output for logging and also provides a mechanism
+  to stream logs over tcp. Setting this to the format `tcp://<host>:<port>` will enable the socket
+  logger for tcp streaming.
+
+* `log.format` can be set to either `text` or `json`. When set to `json` all fields are encapsulated
+  in potentially multi-line `json` fields. This allows for capturing and indexing exceptions as singular fields in
+  a logging search system.
+
+* `log.system-context-file` allows a file containing fields to be set at startup that can be
```
Comment (martint) [inline docs/add_logging.md:55]: log.annotation-file

File: docs/add_logging.md [hunk 4]
```diff
@@ -38,3 +38,45 @@ public class MyClass
...
+  interpolated from environment variables and will be included in all log output (per-line for
+  the `json` formatter and per-file for the `text` formatter).
```
Comment (martint) [inline docs/add_logging.md:57]: both "per line"

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java [hunk 5]
```diff
@@ -3,20 +3,27 @@
...
 
 import java.io.IOException;
 import java.io.StringWriter;
-import java.util.logging.Formatter;
+import java.util.Map;
 import java.util.logging.LogRecord;
 
 public class JsonFormatter
-        extends Formatter
+        extends java.util.logging.Formatter
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/JsonFormatter.java:17]: Why fully qualified?

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java [hunk 6]
```diff
@@ -3,20 +3,27 @@
...
 {
     private static final JsonCodec<JsonRecord> CODEC = new JsonCodecFactory(new ObjectMapperProvider()).jsonCodec(JsonRecord.class);
     private static final JsonFactory jsonFactory = new JsonFactory();
+    private final Map<String, String> logAnnotations;
+
+    public JsonFormatter(Map<String, String> logAnnotations)
+    {
+        this.logAnnotations = logAnnotations == null ? null : ImmutableMap.copyOf(logAnnotations);
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/JsonFormatter.java:25]: Instead of using nulls, why not just use empty maps everywhere. It will make it easier and less error prone.

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java [hunk 7]
```diff
@@ -72,4 +81,9 @@ String minimalJsonErrorLogLine(JsonRecord jsonRecord, Exception exception)
         }
         return stringWriter.append("\n").toString();
     }
+
+    private static String toString(JsonRecord jsonRecord)
+    {
+        return new StringWriter().append(CODEC.toJson(jsonRecord)).append("\n").toString();
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/JsonFormatter.java:87]: Format as:

```
return new StringWriter()
    .append(CODEC.toJson(jsonRecord))
    .append("\n")
    .toString();
```

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java [hunk 8]
```diff
@@ -21,9 +25,10 @@
 import static java.time.temporal.ChronoField.YEAR;
 
 class StaticFormatter
-        extends Formatter
+        extends java.util.logging.Formatter
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/StaticFormatter.java:28]: Why?

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java [hunk 9]
```diff
@@ -66,6 +81,13 @@ public String format(LogRecord record)
                 .append('\t')
                 .append(record.getMessage());
 
+        if (hasLogAnnotations()) {
+            stringWriter.append('\t')
```
Comment (martint) [inline log-manager/src/main/java/io/airlift/log/StaticFormatter.java:85]: The annotations should probably go before the message, which may contain new lines in some cases.

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java (hunk 5 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/JsonFormatter.java:17]: oops, I think because the order I refactored things in, didn't notice it.

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java (hunk 6 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/JsonFormatter.java:25]: I'd need to special case the JsonRecord getter to return null then or else all the json log lines would contain an empty map.

Comment (martint) [reply log-manager/src/main/java/io/airlift/log/JsonFormatter.java:26]: Yeah, but it's better to special-case for JsonRecord when it gets constructed than special case everywhere

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java (hunk 9 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/StaticFormatter.java:86]: My only hesitation with that is if a parser is splitting on tabs, it would mean the ordering of fields would change if the annotations are present or not.

File: docs/add_logging.md [hunk 10]
```diff
@@ -38,3 +38,44 @@ public class MyClass
...
+
+  For example:
+  ```
+  #config.properties
+  log.annotation-file=annotations.properties
+  log.format=json
+  ```
+  ```
+  #annotations.properties
+  hostIp: ${ENV:HOST_IP}
```
Comment (electrum) [inline docs/add_logging.md:66]: I think the `=` syntax is better here, since it's not obvious why these two properties files use a different syntax, and we always use the `=` syntax so most people wouldn't even know this is valid.

File: docs/add_logging.md [hunk 11]
```diff
@@ -38,3 +38,44 @@ public class MyClass
...
+
+* `log.output-file` can be set to the desired output for logging and also provides a mechanism
+  to stream logs over tcp. Setting this to the format `tcp://<host>:<port>` will enable the socket
+  logger for tcp streaming.
+
+* `log.format` can be set to either `text` or `json`. When set to `json` all fields are encapsulated
+  in potentially multi-line `json` fields. This allows for capturing and indexing exceptions as singular fields in
+  a logging search system.
+
+* `log.annontation-file` allows a file containing fields to be set at startup that can be
```
Comment (electrum) [inline docs/add_logging.md:55]: Typo `annontation`

File: docs/add_logging.md [hunk 12]
```diff
@@ -38,3 +38,44 @@ public class MyClass
...
+
+The logging system has several configuration options that are provided to make it easier
+to use in cloud native environments.
+
+* `log.output-file` can be set to the desired output for logging and also provides a mechanism
+  to stream logs over tcp. Setting this to the format `tcp://<host>:<port>` will enable the socket
+  logger for tcp streaming.
+
+* `log.format` can be set to either `text` or `json`. When set to `json` all fields are encapsulated
+  in potentially multi-line `json` fields. This allows for capturing and indexing exceptions as singular fields in
```
Comment (electrum) [inline docs/add_logging.md:52]: Is the multi-line part correct? We don't enable indentation, so the JSON itself shouldn't have newlines, and the string values should have newlines escaped. Maybe rephrase as

> When set to `json`, the log record is formatted as a JSON object, one per line. Any newlines in the field values, such as exception stack traces, will be escaped as normal in JSON object. This allows ...

Comment (electrum) [reply docs/add_logging.md:52]: Nit: write the second `json` instance as JSON without code quotes.

File: log-manager/src/main/java/io/airlift/log/JsonFormatter.java [hunk 13]
```diff
@@ -72,4 +84,12 @@ String minimalJsonErrorLogLine(JsonRecord jsonRecord, Exception exception)
         }
         return stringWriter.append("\n").toString();
     }
+
+    private static String toString(JsonRecord jsonRecord)
+    {
+        return new StringWriter()
```
Comment (electrum) [inline log-manager/src/main/java/io/airlift/log/JsonFormatter.java:90]: No need for `StringWriter`
```java
return CODEC.toJson(jsonRecord) + "\n";

File: log-manager/src/main/java/io/airlift/log/JsonRecord.java [hunk 14]
```diff
@@ -26,14 +28,16 @@ public JsonRecord(
...
+            Throwable throwable,
+            Map<String, String> logAnnotations)
     {
         this.timestamp = requireNonNull(timestamp);
         this.level = requireNonNull(level);
         this.thread = thread;
         this.loggerName = loggerName;
         this.message = message;
         this.throwable = throwable;
+        this.logAnnotations = logAnnotations == null || logAnnotations.isEmpty() ? null : logAnnotations;
```
Comment (electrum) [inline log-manager/src/main/java/io/airlift/log/JsonRecord.java:40]: Nit: add parens around the boolean part

File: log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java [hunk 15]
```diff
@@ -155,4 +156,16 @@ public LoggingConfiguration setLegacyLoggerImplementationEnabled(boolean enabled
...
         return this;
     }
+
+    public String getLogAnnotationFile()
+    {
+        return logAnnotationFile;
+    }
+
+    @Config("log.annotation-file")
+    public LoggingConfiguration setLogAnnotationFile(String logAnnotationFile)
```
Comment (electrum) [inline log-manager/src/main/java/io/airlift/log/LoggingConfiguration.java:166]: We could annotate with `@FileExists`. The config system will validate it exists if not null.

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java [hunk 16]
```diff
@@ -43,6 +50,16 @@
             .appendOffset("+HHMM", "Z")
             .toFormatter(Locale.US);
 
+    public StaticFormatter(Map<String, String> logAnnotations)
+    {
+        this.logAnnotations = ImmutableMap.copyOf(requireNonNull(logAnnotations, "logAnnotations is null"));
+    }
+
+    public StaticFormatter()
```
Comment (electrum) [inline log-manager/src/main/java/io/airlift/log/StaticFormatter.java:58]: Move this above and chain to other constructor
```java
this(ImmutableMap.of());
```
It's preferable to have everything go through a single constructor if possible

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java [hunk 17]
```diff
@@ -62,8 +79,16 @@ public String format(LogRecord record)
                 .append('\t')
                 .append(Thread.currentThread().getName())
                 .append('\t')
-                .append(record.getLoggerName())
-                .append('\t')
+                .append(record.getLoggerName());
+
+        if (!logAnnotations.isEmpty()) {
+            stringWriter.append('\t')
```
Comment (electrum) [inline log-manager/src/main/java/io/airlift/log/StaticFormatter.java:85]: This can be
```java
stringWriter.append('\t');
Joiner.on(",").withKeyValueSeparator("=").appendTo(stringWriter, logAnnotations);
```

File: log-manager/src/test/java/io/airlift/log/TestJsonFormatter.java [hunk 18]
```diff
@@ -66,4 +68,19 @@ public void testLogFormatting()
         assertEquals(jsonMap.get("throwableMessage"), testException.getMessage());
         assertEquals(jsonMap.get("stackTrace"), getStackTraceAsString(testException));
     }
+
+    @Test
+    public void testLogAnnotations()
+    {
+        LogRecord record = new LogRecord(Level.DEBUG.toJulLevel(), "Test Log Message");
+        ImmutableMap<String, String> logAnnotations = ImmutableMap.of("foo", "apple", "bar", "banana");
```
Comment (electrum) [inline log-manager/src/test/java/io/airlift/log/TestJsonFormatter.java:76]: Declare as `Map`

**Review by electrum — APPROVED** [review]
Nit: don't end commit title in period

File: log-manager/src/main/java/io/airlift/log/StaticFormatter.java (hunk 17 above)
Comment (mattstep) [reply log-manager/src/main/java/io/airlift/log/StaticFormatter.java:85]: .appendTo throws IOException, so I just did .join and appended the output

**Comment by mattstep** [comment]
> Nit: don't end commit title in period

fixed
